### PR TITLE
front: add stops in stdcm simulation report sheet v2

### LINKED ISSUE #7790: Add stops in stdcm report sheet
Until then, we had data for departures and arrivals, but not for stops. Now we can display stops on the STDCM summation sheet : 

Image

### PR DESCRIPTION
**Pull request by Akctarus**
closes #7790

### CODE REVIEW

File: front/src/applications/stdcm/components/SimulationReportSheetV2.tsx [hunk 1]
```diff
@@ -290,30 +294,37 @@ const SimulationReportSheetV2 = ({
...
+                      >
                         {
                           // eslint-disable-next-line no-nested-ternary
                           !isFirstStep && !isLastStep
-                            ? // ? step.stop !== '0'
-                              //   ? /* getStopDurationTime(step.stop) */ 'test'
-                              step.time
+                            ? step.duration !== 0
+                              ? getStopDurationTime(step.duration)
+                              : step.time
```
Comment (SharglutDev) [inline front/src/applications/stdcm/components/SimulationReportSheetV2.tsx:319]: Looks like there is not enough vertical padding on top and bottom of this (compared to the mock up) : 

Capture d’écran 2024-06-20 à 09 16 22

File: front/src/applications/stdcm/utils.ts [hunk 2]
```diff
@@ -80,15 +106,27 @@ export function getOperationalPointsWithTimes(simulationReport: SimulationReport
...
+    trainPositions: number[],
+    trainTimes: number[]
+  ): number | null => {
+    const firstIndex = trainPositions.indexOf(position);
+    const lastIndex = trainPositions.lastIndexOf(position);
+    if (firstIndex !== -1 && lastIndex !== -1 && firstIndex !== lastIndex) {
+      return trainTimes[lastIndex] - trainTimes[firstIndex];
+    }
+    return null;
+  };
```
Comment (SharglutDev) [inline front/src/applications/stdcm/utils.ts]: I think this function and all the others defined inside `getOperationalPointsWithTimes` could be moved outside.

Moreover, most of them could benefit from some unit tests. Maybe you can create a new issue to add them in a seperate PR ?

File: front/src/applications/stdcm/utils.ts [hunk 3]
```diff
@@ -98,14 +136,22 @@ export function getOperationalPointsWithTimes(simulationReport: SimulationReport
       departureMinute
     );
 
+    const duration = getDuration(op.positionOnPath, positions, times);
+    const durationInMinutes = duration !== null ? duration / 1000 : 0;
+    const durationToString = secondsToTimeString(durationInMinutes);
```
Comment (SharglutDev) [inline front/src/applications/stdcm/utils.ts]: If you do `/ 1000`, do you rather convert milliseconds to seconds and not to minutes ?

```suggestion
    const durationInSeconds = duration !== null ? duration / 1000 : 0;
    const durationToString = secondsToTimeString(durationInSeconds);
```

File: front/src/applications/stdcm/utils.ts [hunk 4]
```diff
@@ -42,14 +42,40 @@ export function extractSpeedLimit(speedLimitByTag: string): string {
...
+  const seconds = duration % 60;
+  return `${String(minutes).padStart(2, '0')}:${String(seconds).padStart(2, '0')}`;
+}
+
+function timeStringToSeconds(time: string): number {
+  const [minutes, seconds] = time.split(':').map(Number);
+  return minutes * 60 + seconds;
+}
+
+function addTimes(hhmm: string, mmss: string): string {
```
Comment (SharglutDev) [inline front/src/applications/stdcm/utils.ts]:
```suggestion
/**
 * Based on a stop arrival time and a stop duration, calculate the departure time on this stop
 */
function computeStopDepartureTime(hhmm: string, mmss: string): string {
```

File: front/src/applications/stdcm/components/SimulationReportSheetV2.tsx [hunk 5]
```diff
@@ -290,30 +294,37 @@ const SimulationReportSheetV2 = ({
...
+                        }}
+                      >
                         {
                           // eslint-disable-next-line no-nested-ternary
                           !isFirstStep && !isLastStep
-                            ? // ? step.stop !== '0'
-                              //   ? /* getStopDurationTime(step.stop) */ 'test'
-                              step.time
+                            ? step.duration !== 0
+                              ? getStopDurationTime(step.duration)
```
Comment (SharglutDev) [inline front/src/applications/stdcm/components/SimulationReportSheetV2.tsx:318]: You placed this util function in time manipulation but its name is specific to your usecase. Either rename it to a more generic name that can be resued in the app or move the util in utils.ts in stdcm folder.

File: front/src/applications/stdcm/utils.ts [hunk 6]
```diff
@@ -80,15 +106,27 @@ export function getOperationalPointsWithTimes(simulationReport: SimulationReport
...
+  ): string => {
     const index = trainPositions.findIndex((pos) => pos >= trainPosition);
-    if (index === -1) return null;
     const timeInMillis = trainTimes[index];
     const timeInMinutes = Math.floor(timeInMillis / 60000);
     return addMinutesToTime(trainDepartureHour, trainDepartureMinute, timeInMinutes);
   };
 
-  // Map operational points with their positions and times
+  const getDuration = (
```
Comment (SharglutDev) [inline front/src/applications/stdcm/utils.ts]:
```suggestion
  const getStopDurationBetweenToPositions = (
```

**Review by SharglutDev — CHANGES_REQUESTED** [review]
Thank you for the PR. Left some comments.

It would have been nice to have an issue linked to properly know what this PR is supposed to add in the app.

If I check the "asked "path" in the simulation sheet, I see this : 

Capture d’écran 2024-06-20 à 09 26 04

Are you suppose to add a default "reason" for vias ?

**Comment by Akctarus** [comment]
> Are you suppose to add a default "reason" for vias ?

No, because vias can have different reasons. For the moment, we can only determine the reason for stopping the departure and arrival.

File: front/src/applications/stdcm/utils.ts (hunk 6 above)
Comment (Akctarus) [reply front/src/applications/stdcm/utils.ts]: fix

File: front/src/applications/stdcm/components/SimulationReportSheetV2.tsx (hunk 5 above)
Comment (Akctarus) [reply front/src/applications/stdcm/components/SimulationReportSheetV2.tsx:318]: done

File: front/src/applications/stdcm/utils.ts (hunk 4 above)
Comment (Akctarus) [reply front/src/applications/stdcm/utils.ts]: done

File: front/src/applications/stdcm/utils.ts (hunk 3 above)
Comment (Akctarus) [reply front/src/applications/stdcm/utils.ts]: Nice catch ! done

File: front/src/applications/stdcm/utils.ts (hunk 2 above)
Comment (Akctarus) [reply front/src/applications/stdcm/utils.ts]: you're right, done

File: front/src/applications/stdcm/components/SimulationReportSheetV2.tsx (hunk 1 above)
Comment (Akctarus) [reply front/src/applications/stdcm/components/SimulationReportSheetV2.tsx:319]: I tried everything, but if I add 1 more px of margin top and margin bottom, the text inside the box disappears

File: front/src/applications/stdcm/utils.ts (hunk 2 above)
Comment (SharglutDev) [reply front/src/applications/stdcm/utils.ts]: Could you create the enhancement ticket for tests and add it to your board ? So we are sure not to forget about that

**Review by SharglutDev — APPROVED** [review]
Lgtm and tested, good job !

**Pull request merged**